packages/@lwc/template-compiler/src/codegen/helpers.ts

```diff
@@ -102,6 +102,25 @@ export function hasIdAttribute(node: Node): boolean {
     return false;
 }
 
+/**
+ * Returns true if the AST element or any of its descendants uses the `lwc:ref` directive.
+ */
+export function hasRefDirective(node: Node): boolean {
+    if (isBaseElement(node)) {
+        const hasRef = node.directives.some((directive) => directive.name === 'Ref');
+
+        if (hasRef) {
+            return true;
+        }
+    }
+
+    if (isParentNode(node)) {
+        return node.children.some((child) => hasRefDirective(child));
+    }
+
+    return false;
+}
```

Comment:
I'm not super happy with this code – it feels redundant to have to traverse the AST all over again to find `lwc:ref` directives that we already collected earlier. But I wasn't sure of a better way, and I modeled this after the `hasIdAttribute` function above. Happy for suggestions on improvements @jmsjtu @jye-sf

Reply:
Good idea! Done.